qiskit/qasm3/__init__.py

```diff
@@ -53,17 +145,58 @@ def dump(circuit, stream, **kwargs) -> None:
     """Serialize a :class:`~qiskit.circuit.QuantumCircuit` object as a OpenQASM3 stream to file-like
     object.
 
-    .. note::
-
-        This is a quick interface to the main :obj:`.Exporter` interface.  All keyword arguments to
-        this function are inherited from the constructor of that class, and if you have multiple
-        circuits to export, it will be faster to create an :obj:`.Exporter` instance, and use its
-        :obj:`.Exporter.dump` method.
-
     Args:
         circuit (QuantumCircuit): Circuit to serialize.
         stream (TextIOBase): stream-like object to dump the OpenQASM3 serialization
         **kwargs: Arguments for the :obj:`.Exporter` constructor.
 
     """
     Exporter(**kwargs).dump(circuit, stream)
+
+
+@_optionals.HAS_QASM3_IMPORT.require_in_call("loading from OpenQASM 3")
+def load(filename: str):
+    """Load an OpenQASM 3 program from the file ``filename``.
+
+    Args:
+        filename: the filename to load the program from.
+
+    Returns:
+        QuantumCircuit: a circuit representation of the OpenQASM 3 program.
+
+    Raises:
+        QASM3ImporterError: if the OpenQASM 3 file is invalid, or cannot be represented by a
+            :class:`.QuantumCircuit`.
+    """
+
+    import qiskit_qasm3_import
+
+    with open(filename, "r") as fptr:
+        program = fptr.read()
```

Comment:
One random thought here, non-blocking is I wonder if there is a way we can do the parsing from the file directly. Obviously this will require changes in `qiskit_qasm3_import` so it's not a big deal. But, this will result in reading the entire file into memory before the parser can process it, it's generally more efficient for file based access to have the parser read over the file rather than do this. 

That being said I don't expect people to be loading qasm files larger than system memory anytime soon so at worst doing this just adds a bit of overhead.

Reply:
If the file's larger than system memory, then the resultant `QuantumCircuit` will be as well right now, I'm fairly sure.  But yeah, I could add that, but I have to propagate it all the way through to the `openqasm3` base package, which is the component that actually calls ANTLR, which at the moment only allows string input.

Reply:
Semi-related on this topic: right now, I'm not keen on accepting an opened file-like stream (as opposed to a filename) in `load`, because if we decide to replace the internals with something compiled, it's much harder / less efficient to handle a Python file handle, rather than passing the filename across the API boundary and letting the compiled component open the stream through its own means.

Reply:
Yeah, I'm fine with a filename input only. I was more just concerned with reading the full file into a string before parsing it. I agree with a compiled language parser it'll be easier and more performant to just open the file in the lower level language directly for parsing.

Reply:
Even if we read the whole file into memory now, we can read the minimum needed for translating each piece (or buffering, etc.) later without changing the API. That is you still pass a file name. This doesn't preclude expanding the API later to take a stream. It's pretty clear that no one suggested holding this up while everything down to the OpenQASM parser is modified. But it seems like there is no disadvantage at all with respect to this question in merging this as is.